pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,201 @@
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+)
+
+
+var UnixErrorLogPath = "/opt/aws/aws-otel-collector/logs/aws-otel-collector-error.log"
+
+var WindowsErrorLogPath = "C:\\ProgramData\\Amazon\\AWSOTelCollector\\Logs\\aws-otel-collector-error.log"
+
+var errorLogfile = getErrorLogFilePath()
+
+var errorLogMaxAge = 10
+
+var queue = list.New()
+var errorLogChannel = make(chan zapcore.Entry, 10)
+var ticker = time.NewTicker(10 * time.Minute)
+var quit = make(chan struct{})
+var kilobyte = 1024
+var errorFile *os.File
+
+// WriteError func could write the new error log message into file
+func WriteError(newErrorLog zapcore.Entry, errorLogfile string) (n int, err error) {
+	var byteNum int
+	if errorFile == nil {
+		if err = openExistingOrNewErrorFile(); err != nil {
+			return 0, err
+		}
+	}
+	err = errorFile.Truncate(0)
+	if err != nil {
+		return byteNum, err
+	}
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	var backupQueue = list.New()
+	for queue.Len() > 0 {
+		e := queue.Front().Value.(zapcore.Entry)
+		if e.Message != newErrorLog.Message {
+			n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				return byteNum, err
+			}
+			byteNum += n
+			backupQueue.PushBack(e)
+		}
+		queue.Remove(queue.Front())
+	}
+	backupQueue.PushBack(newErrorLog)
+	queue = backupQueue
+	n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newErrorLog.Time, newErrorLog.Level, newErrorLog.Caller, newErrorLog.Message)))
+	if err != nil {
+		return byteNum, err
+	}
+	byteNum += n
+	if byteNum >= kilobyte {
+		byteNum, err = rotate(byteNum, errorLogfile)
+		if err != nil {
+			return byteNum, err
+		}
+	}
+	return byteNum, err
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func rotate(byteNum int, errorLogfile string) (n int, err error) {
+	file, err := os.OpenFile(errorLogfile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		return byteNum, err
+	}
+	err = file.Truncate(0)
+	if err != nil {
+		return byteNum, err
+	}
+	errorNum := queue.Len()
+	for i := 0; i < errorNum; i++ {
+		currentError := queue.Front().Value.(zapcore.Entry)
+		currentLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+			currentError.Time, currentError.Level, currentError.Caller, currentError.Message))
+		if byteNum > kilobyte {
+			byteNum -= len(currentLog)
+			queue.Remove(queue.Front())
+		} else {
+			_, err = file.Write(currentLog)
+			if err != nil {
+				return byteNum, err
+			}
+		}
+
+	}
+	return byteNum, err
+}
+
+var RotateError = make(chan error)
+
+// RotateErrorLog func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func RotateErrorLog() {
```

Comment:
I still don't think this is an ideal function name. This function might as well be called `ECSErrorReporter` or something. Its like the main function for this feature. It does more than just rotation.

Reply:
good idea, will change that